bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go

```diff
@@ -495,6 +504,9 @@ func (r *KubeadmConfigReconciler) handleClusterNotInitialized(ctx context.Contex
 		return ctrl.Result{}, err
 	}
 
+	// Update the certificate expiration time in the config.
+	r.addCertificateExpiryAnnotation(scope.Config)
```

Comment:
Is this reentrant? What happens if the controller fails after l.502? (similar for the other case)

Reply:
As far as I remember the controller keeps re-trying until we set bootstrap data, at the end of this func, but this probably requires some more investigations. However it doesn't seems to me that this addition makes things worst in the case we might have some re-entrancy problems

Reply:
@ykakarap Maybe you can take a quick (slightly) deeper look

Reply:
If the controller fails at l.502 or any of the previous places the `KubeadmConfig.Status.Ready` will not be set to true. The controller keeps retrying to create the bootstrap data until `.Status.Ready` is true. 
The `.Status.Ready` will be true only if the bootstrap data is successfully strored. 
Therefore it is okay if any of the previous steps error as the controller will keep re-trying untill the bootstrap data is successfully created and stored. 

Note: It doent really matter when parts of the the boostrap data (CA certificates) are created. We only care about when the final bootstrap data secret is created. This is becasue the machine does not do anything untill the full bootstrap data secret is available. The certificates on the machine that expire in one year (the ones that need to be rotated) are created when `kubeadm (init|join)` runs on the underlying machine. This only happens after the bootstrap data is available.

Reply:
Sounds perfect. Thank you!